File: 31-next-permutation.js
const swap = (nums, index1, index2) => {
  let temp = nums[index1];
  nums[index1] = nums[index2];
  nums[index2] = temp;
};

var nextPermutation = function (nums) {
  if (nums.length <= 1) {
    return nums;
  }

  let i = nums.length - 2;

  for (; i >= 0; i--) {
    if (nums[i] < nums[i + 1]) {
      break;
    }
  }

  if (i >= 0) {
    let j = nums.length - 1;

    for (; j > i; j--) {
      if (nums[j] > nums[i]) {
        break;
      }
    }

    swap(nums, i, j);
  }

  // Reverse the subarray to the right of i
  for (let start = i + 1, end = nums.length - 1; start < end; start++, end--) {
    swap(nums, start, end);
  }

  return nums;
};

const inputArray = [1, 3, 2];
nextPermutation(inputArray);
console.log(inputArray);

File: 31-next-permutation.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void print(vector<int> nums)
{
    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << " ";
    }
    cout << endl;
}

class Solution
{
public:
    void nextPermutation(vector<int> &nums)
    {
        if (nums.size() == 2 && nums[0] != nums[1] && nums.size() > 1)
        {
            swap(nums[0], nums[1]);
        }
        else if (nums.size() > 2)
        {
            int index = -1;

            for (int i = nums.size() - 2; i >= 0; i--)
            {
                if (nums[i] < nums[i + 1])
                {
                    index = i;
                    break;
                }
            }

            if (index == -1)
            {
                reverse(nums.begin(), nums.end());
            }
            else
            {
                for (int i = nums.size() - 1; i > index; i--)
                {
                    if (nums[i] > nums[index])
                    {
                        swap(nums[i], nums[index]);
                        break;
                    }
                }
                reverse(nums.begin() + index + 1, nums.end());
            }
        }
        print(nums);
    }
};

int main()
{
    vector<int> nums = {1};
    Solution s;
    s.nextPermutation(nums);
    return 0;
}